### PR TITLE
Update to statistics class

### PR DESCRIPTION
**Pull request by Kopkins**
Moved statistics class to its own file and added functionality to write the statistics to a DB file at /data/valhalla/statistics.sqlite.

### CODE REVIEW

File: src/mjolnir/graphvalidator.cc [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
-#include "valhalla/mjolnir/graphtilebuilder.h"
+#include "mjolnir/graphtilebuilder.h"
+#include "mjolnir/statistics.h"
 
 #include <valhalla/midgard/logging.h>
 
 #include <ostream>
-#include <set>
 #include <boost/format.hpp>
+#include <boost/filesystem/operations.hpp>
```
Comment (kevinkreiser) [inline src/mjolnir/graphvalidator.cc]: not needed i think

File: src/mjolnir/statistics.cc [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+using namespace valhalla::mjolnir;
+
+namespace valhalla {
+namespace mjolnir {
+
+validator_stats::validator_stats ()
+  : tile_maps(), country_maps(), iso_codes(), tile_ids(), dupcounts(3), densities(3) { }
+
+void validator_stats::add_tile_road (const uint32_t& tile_id, const RoadClass& rclass, float length) {
+  this->tile_ids.insert(tile_id);
```
Comment (kevinkreiser) [inline src/mjolnir/statistics.cc]: we haven't been using the `this` keyword at all in any of the other classes. i am personally not opposed to it (i secretely like it) but maybe for the sake of consistency we should remove it

File: src/mjolnir/graphvalidator.cc (hunk 1 above)
Comment (Kopkins) [inline src/mjolnir/graphvalidator.cc]: was leftover from when the whole class was in that file.

**Comment by dnesbitt61** [comment]
:shipit:

File: src/mjolnir/statistics.cc [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    LOG_ERROR("cannot open " + database);
+    sqlite3_close(db_handle);
+    db_handle = NULL;
+    return;
+  }
+
+  // Create table for tiles
+  sql = "CREATE TABLE tiledata (";
+  sql += "tileid INTEGER,";
+  sql += "motorway FLOAT,";
```
Comment (gknisely) [inline src/mjolnir/statistics.cc]: Nothing major....but I usually follow the data types defined here:
https://www.sqlite.org/datatype3.html

Each column in an SQLite 3 database is assigned one of the following type affinities:

TEXT
NUMERIC
INTEGER
REAL
NONE

File: src/mjolnir/statistics.cc [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+  ret = sqlite3_exec(db_handle, "BEGIN", NULL, NULL, &err_msg);
+  if (ret != SQLITE_OK) {
+    LOG_ERROR("Error: " + std::string(err_msg));
+    sqlite3_free(err_msg);
+    sqlite3_close(db_handle);
+  }
+  sql = "INSERT INTO tiledata (tileid, motorway, trunk, pmary, secondary, tertiary, unclassified, residential, serviceother) ";
+  sql += "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)";
+  ret = sqlite3_prepare_v2(db_handle, sql.c_str(), strlen (sql.c_str()), &stmt, NULL);
+  if (ret != SQLITE_OK) {
```
Comment (gknisely) [inline src/mjolnir/statistics.cc]: Is ok to continue here or should we fail?

File: src/mjolnir/statistics.cc [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    LOG_ERROR("Error: " + std::string(err_msg));
+    sqlite3_free(err_msg);
+    sqlite3_close(db_handle);
+  }
+  sql = "INSERT INTO countrydata (isocode, motorway, trunk, pmary, secondary, tertiary, unclassified, residential, serviceother) ";
+  sql += "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)";
+  ret = sqlite3_prepare_v2(db_handle, sql.c_str(), strlen (sql.c_str()), &stmt, NULL);
+  if (ret != SQLITE_OK) {
+    LOG_ERROR("SQL error: " + sql);
```
Comment (gknisely) [inline src/mjolnir/statistics.cc]: Same...should we stop?

**Comment by gknisely** [comment]
That is all I saw.  :+1:

File: src/mjolnir/statistics.cc (hunk 4 above)
Comment (Kopkins) [inline src/mjolnir/statistics.cc]: If we should return on error, we should add it to pbfadminbuilder too. Line 325

**Comment by gknisely** [comment]
:shipit:

File: src/mjolnir/statistics.cc (hunk 5 above)
Comment (dnesbitt61) [inline src/mjolnir/statistics.cc]: I would never exit due to not being able to generate statistics. I can't imagine doing the whole planet successfully only to exit here. Log it is fine.

Comment (Kopkins) [inline src/mjolnir/statistics.cc]: At the point where this code would execute the only remaining this is to write to the DB. Maybe we can review this tomorrow

**Pull request merged**